pkg/gadgets/trace/tcpdrop/tracer/tracer.go

```diff
@@ -0,0 +1,246 @@
+
+	if err := spec.LoadAndAssign(&t.objs, &opts); err != nil {
+		var errVerifier *ebpf.VerifierError
+		if errors.As(err, &errVerifier) {
+			fmt.Printf("Verifier error: %+v\n",
+				errVerifier)
+		}
+		return fmt.Errorf("failed to load ebpf program: %w", err)
+	}
+
+	t.kfreeSkbLink, err = link.Tracepoint("skb", "kfree_skb", t.objs.IgTcpdrop, nil)
+	if err != nil {
+		return fmt.Errorf("error opening tracepoint kfree_skb: %w", err)
+	}
+
+	reader, err := perf.NewReader(t.objs.tcpdropMaps.Events, gadgets.PerfBufferPages*os.Getpagesize())
+	if err != nil {
+		return fmt.Errorf("error creating perf ring buffer: %w", err)
+	}
+	t.reader = reader
+
+	return nil
+}
+
+func (t *Tracer) run() {
+	for {
+		record, err := t.reader.Read()
+		if err != nil {
+			if errors.Is(err, perf.ErrClosed) {
+				// nothing to do, we're done
+				return
+			}
+
+			msg := fmt.Sprintf("Error reading perf ring buffer: %s", err)
+			t.eventCallback(types.Base(eventtypes.Err(msg)))
+			return
+		}
+
+		if record.LostSamples > 0 {
+			msg := fmt.Sprintf("lost %d samples", record.LostSamples)
+			t.eventCallback(types.Base(eventtypes.Warn(msg)))
+			continue
+		}
+
+		bpfEvent := (*tcpdropEvent)(unsafe.Pointer(&record.RawSample[0]))
+
+		reason := lookupDropReason(int(bpfEvent.Reason))
+
+		event := types.Event{
+			Event: eventtypes.Event{
+				Type:      eventtypes.NORMAL,
+				Timestamp: gadgets.WallTimeFromBootTime(bpfEvent.Timestamp),
+			},
+			WithMountNsID: eventtypes.WithMountNsID{MountNsID: bpfEvent.MntnsId},
+			Pid:           bpfEvent.Pid,
+			Comm:          gadgets.FromCString(bpfEvent.Comm[:]),
+			Dport:         gadgets.Htons(bpfEvent.Dport),
+			Sport:         gadgets.Htons(bpfEvent.Sport),
+			State:         tcpbits.TCPState(bpfEvent.State),
+			Tcpflags:      tcpbits.TCPFlags(bpfEvent.Tcpflags),
+			Reason:        reason,
+		}
+
+		if bpfEvent.Af == unix.AF_INET {
+			event.IPVersion = 4
+		} else if bpfEvent.Af == unix.AF_INET6 {
+			event.IPVersion = 6
+		}
+
+		event.Saddr = gadgets.IPStringFromBytes(bpfEvent.Saddr, event.IPVersion)
+		event.Daddr = gadgets.IPStringFromBytes(bpfEvent.Daddr, event.IPVersion)
+
+		if t.enricher != nil {
+			t.enricher.EnrichByMntNs(&event.CommonData, event.MountNsID)
+		}
+
+		t.eventCallback(&event)
+	}
+}
+
+var (
+	btfDetectDropReason sync.Once
+	dropReasons         = map[int]string{}
+	dropReasonsErr      error
+)
+
+func lookupDropReason(reason int) string {
+	btfDetectDropReason.Do(func() {
+		btfSpec, err := btf.LoadKernelSpec()
+		if err != nil {
+			dropReasonsErr = err
+			return
+		}
+
+		enum := &btf.Enum{}
+		err = btfSpec.TypeByName("skb_drop_reason", &enum)
+		if err != nil {
+			dropReasonsErr = err
+			return
+		}
+		for _, v := range enum.Values {
+			dropReasons[int(v.Value)] = v.Name
+		}
+	})
+
+	if dropReasonsErr != nil {
+		return dropReasonsErr.Error()
+	}
+	if ret, ok := dropReasons[reason]; ok {
+		ret = strings.TrimPrefix(ret, "SKB_DROP_REASON_")
+		ret = strings.TrimPrefix(ret, "SKB_")
```

Comment:
nit: do this above in L195 to avoid having to make this every time to each event.